Improve reusable block creation flow

noisysocks opened the pull request:
## Description

Closes #4527.

Modifies the flow to create a reusable block to look like this:

1) The user selects a block and clicks _Convert to Reusable Block_.

<img width="713" alt="screen shot 2018-02-28 at 10 28 02" src="https://user-images.githubusercontent.com/612155/36761290-74e7dc26-1c72-11e8-8797-518a2ed21f66.png">

2) The reusable block is created and is immediately in editing mode. **The title of the block is selected, allowing the user to start typing the title straight away.** If the user abandons the flow at this point, the new block will be titled  _Untitled block_.

<img width="709" alt="screen shot 2018-02-28 at 10 28 18" src="https://user-images.githubusercontent.com/612155/36761296-7ad085a2-1c72-11e8-9526-62087b396bd3.png">

3) The user clicks _Save_ to rename their reusable block.

<img width="697" alt="screen shot 2018-02-28 at 10 28 41" src="https://user-images.githubusercontent.com/612155/36761338-aca7e458-1c72-11e8-8cb4-d27101fdab7e.png">

This removes two steps from the flow of creating a reusable block.

<!-- Please describe your changes -->

## Testing

1. Create a block
2. Convert it to a reusable block. Confirm that the title input is selected straight away
3. Test creating and inserting reusable blocks

noisysocks committed:
Improve reusable block creation flow

When a reusable block is created, immediately put it into editing mode
and select the title <input>. This reduces the amount of steps required
to create a reusable block.

aduth commented on blocks/library/block/edit-panel/index.js:
```diff
@@ -10,66 +11,97 @@ import { keycodes } from '@wordpress/utils';
  */
 import './style.scss';
 
+/**
```
FWIW I don't think it's so bad in this case, because it's fairly obvious what `ESCAPE` is, but I dislike the idea of a "Module constants" docblock in general, because it normalizes the idea that we don't need to document variables we define, and can instead apply a blanket grouping to them.

noisysocks replied:
I'll leave this for now since we do this in a few places and would prefer we're somewhat consistent. It might be worth going through and auditing our usage of this pattern, since you raise a really good point.

aduth commented on blocks/library/block/edit-panel/index.js:
```diff
@@ -10,66 +11,97 @@ import { keycodes } from '@wordpress/utils';
+		this.titleRef = ref;
+	}
+
+	componentDidMount() {
```
Conventionally like to see all component lifecycle following the constructor, with no functions in between.

noisysocks replied:
👌 I've moved it in c7e444f64c3df8edc56ae9b309ee98902e3acc3c.

aduth commented on blocks/library/block/index.js:
```diff
@@ -27,8 +27,10 @@ class ReusableBlockEdit extends Component {
 		this.setTitle = this.setTitle.bind( this );
 		this.updateReusableBlock = this.updateReusableBlock.bind( this );
 
+		const { reusableBlock } = this.props;
```
I'd thought this not to work in constructor, since `props` is otherwise passed as the first argument of the constructor function. Not sure if there's a strong consensus on whether this should be discouraged though.

noisysocks replied:
If nothing else, it's simpler! Fixed in 5af814ea52a4fc78c8398f24cdec9e4883e1d457.

aduth commented on blocks/library/block/edit-panel/index.js:
```diff
@@ -10,66 +11,97 @@ import { keycodes } from '@wordpress/utils';
-		( ! isEditing && ! isSaving ) && (
-			<div key="view" className="reusable-block-edit-panel">
-				<span className="reusable-block-edit-panel__info">
-					<b>{ title }</b>
```
Of course not relevant to this pull request, but was `<b>` chosen for a particular reason?

>Do not confuse the <b> element with the <strong>, <em>, or <mark> elements. The <strong> element represents text of certain importance, <em> puts some emphasis on the text and the <mark> element represents text of certain relevance. The <b> element doesn't convey such special semantic information; use it only when no others fit.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/b

Further, do we need _both_ the `span` and the `b`, or could we collapse to eliminate one or the other (probably just replacing `span` with desired `inline` tag).

noisysocks replied:
Eliminated the `<span>` in 3b92ad56e11687850d0efb378bdda1437e0be90c. I like the `<b>` in lieu of a `<span>` since it denotes that there's something special about this text to the developer who is reading the JSX.

noisysocks committed:
Re-arrange ReusableBlockEditPanel methods

Move bindTitleRef so that the React lifecycle methods immediately follow
the constructor.

noisysocks committed:
Don't access this.props in the ReusableBlockEdit constructor

There's no need, as we're passed props as the constructor's first
argument.

noisysocks committed:
Eliminate unnecessary <span>

Simplify the edit panel DOM tree by replacing <span><b /></span> with
just the single <b>.
